aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -294,6 +341,113 @@ void multi_tensor_apply(
   }
 }
 
+template <int depth, typename T, typename... ArgTypes>
+void multi_tensor_apply(
+    std::vector<std::vector<at::Tensor>>& tensor_lists,
+    T callable,
+    ArgTypes... args) {
+  // Note: [static arg vs. dynamic arg]
+  // Due to the dynamic nature of the workload, the kernel arguments aren't
+  // guaranteed to fit in the static 4kb kernel argument memory. Previously
+  // with the apex implementation, we overcame this limitation by dividing a
+  // multi_tensor_apply workload into multiple kernel launches. However, this
+  // led to low sustained occupancy, affecting the performance of memory bound
+  // ops.
+  //
+  // Based on the observation that the kernel argument memory limitation
+  // doesn't correlate well with available SM resources, we have adopted a
+  // different approach. When the kernel arguments fit into the static kernel
+  // argument memory, we use this memory to transfer the arguments. Conversely,
+  // when the kernel arguments don't fit into the static kernel argument
+  // memory, instead of sacrificing sustained occupancy, we use a page-locked
+  // cudaMemcpyAsync to transfer the arguments, then perform the entire
+  // workload in a single kernel.
+  if (can_use_static_tensor_list_meta(tensor_lists, depth)) {
+    multi_tensor_apply_static<depth, T, ArgTypes...>(
+        tensor_lists, callable, args...);
+    return;
+  }
+
+  TORCH_CHECK(
+      tensor_lists.size() == depth,
+      "Number of tensor lists has to match the depth.");
+  const size_t n_tensors = tensor_lists[0].size();
+
+  TORCH_CHECK(
+      n_tensors < static_cast<size_t>(std::numeric_limits<int>::max()),
+      "multi_tensor_apply: exceeded the maximum supported number of tensors: ",
+      n_tensors,
+      " > ",
+      std::numeric_limits<int>::max());
+
+  std::vector<const void*> addresses[depth];
+  std::vector<int64_t> numel_for_tensor;
+  std::vector<uint32_t> block_to_tensor;
+  std::vector<uint32_t> block_to_chunk;
+
+  for (int d = 0; d < depth; ++d) {
+    addresses[d].reserve(n_tensors);
+  }
+  numel_for_tensor.reserve(n_tensors);
+  block_to_tensor.reserve(n_tensors); // reserve for lowerbound
+  block_to_chunk.reserve(n_tensors); // reserve for lowerbound
+
+  for (size_t t = 0; t < n_tensors; t++) {
+    // short-circuit to avoid adding empty tensors to tensorListMeta
+    if (tensor_lists[0][t].numel() == 0) {
+      continue;
+    }
+    numel_for_tensor.push_back(tensor_lists[0][t].numel());
+    for (int d = 0; d < depth; d++) {
+      addresses[d].push_back(tensor_lists[d][t].const_data_ptr());
+    }
+
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
+    // Using float32 as example, hitting this limit requires rougly 2^32 *
```

Comment:
```suggestion
    // Using float32 as example, hitting this limit requires roughly 2^32 *
```
another nit

But real q: do we have precedence of setting a "max number of tensors dealt with at time" anywhere else?

Reply:
I've not seen any intentional cases. Maybe for peace of mind we should just switch to `size_t` here, since we are dynamically allocating kernel arguments and the extra size wouldn't really matter perf-wise.

Reply:
Does `size_t` have the same unstandardized bitwidth problem as `int` across different OS's? If not, then sounds good to me.